### DISCUSSION TITLE
UEFI: Block I/O Protocol Availability

### DISCUSSION THREAD

**Discussion by kennystrawnmusic** [phil-opp/blog_os#1134] [Questions] [answered]
Is the Block I/O protocol accessible after boot services are exited, seeing as though the Graphics Output Protocol definitely is? Or do I have to load some external binary (e.g. [an init process](https://github.com/kennystrawnmusic/cryptos/blob/ac254bc9163b3fa5f176a2b575f2b1fec8a62fda/src/main.rs#L67)) from the disk while boot services are still active and store it in a global for later execution if I want to use the Block I/O protocol to load it?

**Comment by bjorn3** [comment, marked as answer]
I believe the Graphics Output Protocol isn't actually available after exiting boot services. Or to be precise you can't change any settings of it anymore. Only write to a framebuffer that has already been setup before exiting boot services. As for block IO I expect it to be entirely unavailable after exiting boot services. Before exiting boot services UEFI is in control of all devices and it can for example change page tables as necessary. After exiting boot services the OS is in full control of all devices and UEFI can't change page tables anymore. For all UEFI knows you have already configured the AHCI/NVMe/USB controller backing the block device and messing around with it would cause problems. As such I don't expect block IO to be working after boot services are exited.

**Reply by kennystrawnmusic** [reply]
Thanks, that's what I figured.

**Reply by kennystrawnmusic** [reply]
Just makes one wonder how exactly it's even possible for microkernel systems (e.g. Redox) to load user-mode drivers from disk in order to run them if they can't even access the disk to load the drivers from.

**Reply by bjorn3** [reply]
Microkernels generally embed an init ramdisk inside the kernel image. This init ramdisk contains the first program to start and all essential drivers (disk driver, filesystem driver, ...) which will be started by this initial program.
